native-installers/debian/dependencies/files/opt/sigma/E/DOC/man/eprover.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.37.1.
.TH E "1" "July 2013" "E 1.8-001 Gopaldhara" "User Commands"
.SH NAME
E \- manual page for E 1.8-001 Gopaldhara
.SH SYNOPSIS
.B eprover
[\fIoptions\fR] [\fIfiles\fR]
.SH DESCRIPTION
E 1.8\-001 "Gopaldhara"
.PP
Read a set of first\-order clauses and formulae and try to refute it.
.SH OPTIONS

.HP
\fB\-h\fR
.HP
\fB\-\-help\fR
.IP
Print a short description of program usage and options.
.HP
\fB\-V\fR
.HP
\fB\-\-version\fR
.IP
Print the version number of the prover. Please include this with all bug
reports (if any).
.HP
\fB\-v\fR
.HP
\fB\-\-verbose[=\fR<arg>]
.IP
Verbose comments on the progress of the program. This differs from the
output level (below) in that technical information is printed to stderr,
while the output level determines which logical manipulations of the
clauses are printed to stdout. The short form or the long form without
the optional argument is equivalent to \fB\-\-verbose\fR=\fI1\fR.
.HP
\fB\-o\fR <arg>
.HP
\fB\-\-output\-file=\fR<arg>
.IP
Redirect output into the named file.
.HP
\fB\-s\fR
.HP
\fB\-\-silent\fR
.IP
Equivalent to \fB\-\-output\-level\fR=\fI0\fR.
.HP
\fB\-l\fR <arg>
.HP
\fB\-\-output\-level=\fR<arg>
.IP
Select an output level, greater values imply more verbose output. Level 0
produces nearly no output, level 1 will output each clause as it is
processed, level 2 will output generating inferences, level 3 will give a
full protocol including rewrite steps and level 4 will include some
internal clause renamings. Levels >= 2 also imply PCL2 or TSTP formats
(which can be post\-processed with suitable tools).
.HP
\fB\-p\fR
.HP
\fB\-\-proof\-object[=\fR<arg>]
.IP
Generate (and print, in case of success) an internal proof object. Level
0 will not build a proof object, level 1 will build a simple, compact
proof object that only contains inference rules and dependencies, level 2
will build a proof object where inferences are unambiguously described by
giving inference positions, and level 3 will expand this to a proof
object where all intermediate results are explicit. This feature is under
development, so far only level 0 and 1 are operational. The short form or
the long form without the optional argument is equivalent to
\fB\-\-proof\-object\fR=\fI1\fR.
.HP
\fB\-\-pcl\-terms\-compressed\fR
.IP
Print terms in the PCL output in shared representation.
.HP
\fB\-\-pcl\-compact\fR
.IP
Print PCL steps without additional spaces for formatting (safes disk
space for large protocols).
.HP
\fB\-\-pcl\-shell\-level[=\fR<arg>]
.IP
Determines level to which clauses and formulas are suppressed in the
output. Level 0 will print all, level 1 will only print initial
clauses/formulas, level 2 will print no clauses or axioms. All levels
will still print the dependency graph. The option without the optional
argument is equivalent to \fB\-\-pcl\-shell\-level\fR=\fI1\fR.
.HP
\fB\-\-print\-statistics\fR
.IP
Print the inference statistics (only relevant for output level
0,otherwise they are printed automatically.
.HP
\fB\-0\fR
.HP
\fB\-\-print\-detailed\-statistics\fR
.IP
Print data about the proof state that is potentially expensive to
collect. Includes number of term cells and number of rewrite steps.
.HP
\fB\-S\fR
.HP
\fB\-\-print\-saturated[=\fR<arg>]
.IP
Print the (semi\-) saturated clause sets after terminating the saturation
process. The argument given describes which parts should be printed in
which order. Legal characters are 'eigEIGaA', standing for processed
positive units, processed negative units, processed non\-units,
unprocessed positive units, unprocessed negative units, unprocessed
non\-units, and two types of additional equality axioms, respectively.
Equality axioms will only be printed if the original specification
contained real equality. In this case, 'a' requests axioms in which a
separate substitutivity axiom is given for each argument position of a
function or predicate symbol, while 'A' requests a single substitutivity
axiom (covering all positions) for each symbol. The short form or the
long form without the optional argument is equivalent to
\fB\-\-print\-saturated\fR=\fIeigEIG\fR.
.HP
\fB\-\-print\-sat\-info\fR
.IP
Print additional information (clause number, weight, etc) as a comment
for clauses from the semi\-saturated end system.
.HP
\fB\-\-filter\-saturated[=\fR<arg>]
.TP
Filter the
(semi\-) saturated clause sets after terminating the
.IP
saturation process. The argument is a string describing which operations
to take (and in which order). Options are 'u' (remove all clauses with
more than one literal), 'c' (delete all but one copy of identical
clauses, 'n', 'r', 'f' (forward contraction, unit\-subsumption only, no
rewriting, rewriting with rules only, full rewriting, respectively), and
\&'N', 'R' and 'F' (as their lower case counterparts, but with
non\-unit\-subsumption enabled as well). The option without the optional
argument is equivalent to \fB\-\-filter\-saturated\fR=\fIFc\fR.
.HP
\fB\-\-prune\fR
.IP
Stop after relevancy pruning, SInE pruning, and output of the initial
clause\- and formula set. This will automatically set output level to 4 so
that the pruned problem specification is printed. Note that the desired
pruning methods must still be specified (e.g. '\-\-sine=Auto'
.HP
\fB\-\-cnf\fR
.IP
Convert the input problem into clause normal form and print it. This is
(nearly) equivalent to '\-\-print\-saturated=eigEIG
\fB\-\-processed\-clauses\-limit\fR=\fI0\fR' and will by default perform some usually
useful simplifications. You can additionally specify e.g.
\&'\-\-no\-preprocessing' if you want just the result of CNF translation.
.HP
\fB\-\-print\-pid\fR
.IP
Print the process id of the prover as a comment after option processing.
.HP
\fB\-\-print\-version\fR
.IP
Print the version number of the prover as a comment after option
processing. Note that unlike \fB\-version\fR, the prover will not terminate, but
proceed normally.
.HP
\fB\-\-error\-on\-empty\fR
.IP
Return with an error code if the input file contains no clauses.
Formally, the empty clause set (as an empty conjunction of clauses) is
trivially satisfiable, and E will treat any empty input set as
satisfiable. However, in composite systems this is more often a sign that
something went wrong. Use this option to catch such bugs.
.HP
\fB\-m\fR <arg>
.HP
\fB\-\-memory\-limit=\fR<arg>
.IP
Limit the memory the prover may use. The argument is the allowed amount
of memory in MB. If you use the argument 'Auto', the system will try to
figure out the amount of physical memory of your machine and claim most
of it. This option may not work everywhere, due to broken and/or strange
behaviour of setrlimit() in some UNIX implementations, and due to the
fact that I know of no portable way to figure out the physical memory in
a machine. Both the option and the 'Auto' version do work under all
tested versions of Solaris and GNU/Linux. Due to problems with limit data
types, it is currently impossible to set a limit of more than 2 GB (2048
MB).
.HP
\fB\-\-cpu\-limit[=\fR<arg>]
.IP
Limit the cpu time the prover should run. The optional argument is the
CPU time in seconds. The prover will terminate immediately after reaching
the time limit, regardless of internal state. This option may not work
everywhere, due to broken and/or strange behaviour of setrlimit() in some
UNIX implementations. It does work under all tested versions of Solaris,
HP\-UX, MacOS\-X, and GNU/Linux. As a side effect, this option will inhibit
core file writing. Please note that if you use both \fB\-\-cpu\-limit\fR and
\fB\-\-soft\-cpu\-limit\fR, the soft limit has to be smaller than the hard limit to
have any effect.  The option without the optional argument is equivalent
to \fB\-\-cpu\-limit\fR=\fI300\fR.
.HP
\fB\-\-soft\-cpu\-limit[=\fR<arg>]
.IP
Limit the cpu time the prover should spend in the main saturation phase.
The prover will then terminate gracefully, i.e. it will perform
post\-processing, filtering and printing of unprocessed clauses, if these
options are selected. Note that for some filtering options (in particular
those which perform full subsumption), the post\-processing time may well
be larger than the saturation time. This option is particularly useful if
you want to use E as a preprocessor or lemma generator in a larger
system. The option without the optional argument is equivalent to
\fB\-\-soft\-cpu\-limit\fR=\fI290\fR.
.HP
\fB\-R\fR
.HP
\fB\-\-resources\-info\fR
.IP
Give some information about the resources used by the prover. You will
usually get CPU time information. On systems returning more information
with the rusage() system call, you will also get information about memory
consumption.
.HP
\fB\-C\fR <arg>
.HP
\fB\-\-processed\-clauses\-limit=\fR<arg>
.IP
Set the maximal number of clauses to process (i.e. the number of
traversals of the main\-loop).
.HP
\fB\-\-answers[=\fR<arg>]
.IP
Set the maximal number of answers to print for existentially quantified
questions. Without this option, the prover terminates after the first
answer found. If the value is different from 1, the prover is no longer
guaranteed to terminate, even if there is a finite number of answers. The
option without the optional argument is equivalent to
\fB\-\-answers\fR=\fI2147483647\fR.
.HP
\fB\-\-conjectures\-are\-questions\fR
.IP
Treat all conjectures as questions to be answered. This is a wart
necessary because CASC\-J6 has categories requiring answers, but does not
yet support the 'question' type for formulas.
.HP
\fB\-P\fR <arg>
.HP
\fB\-\-processed\-set\-limit=\fR<arg>
.IP
Set the maximal size of the set of processed clauses. This differs from
the previous option in that redundant and back\-simplified processed
clauses are not counted.
.HP
\fB\-U\fR <arg>
.HP
\fB\-\-unprocessed\-limit=\fR<arg>
.IP
Set the maximal size of the set of unprocessed clauses. This is a
termination condition, not something to use to control the deletion of
bad clauses. Compare \fB\-\-delete\-bad\-limit\fR.
.HP
\fB\-T\fR <arg>
.HP
\fB\-\-total\-clause\-set\-limit=\fR<arg>
.IP
Set the maximal size of the set of all clauses. See previous option.
.HP
\fB\-n\fR
.HP
\fB\-\-eqn\-no\-infix\fR
.IP
In LOP, print equations in prefix notation equal(x,y).
.HP
\fB\-e\fR
.HP
\fB\-\-full\-equational\-rep\fR
.IP
In LOP. print all literals as equations, even non\-equational ones.
.HP
\fB\-\-tptp\-in\fR
.IP
Parse TPTP\-2 format instead of E\-LOP (but note that includes are handled
according to TPTP\-3 semantics).
.HP
\fB\-\-tptp\-out\fR
.IP
Print TPTP format instead of E\-LOP. Implies \fB\-\-eqn\-no\-infix\fR and will
ignore \fB\-\-full\-equational\-rep\fR.
.HP
\fB\-\-tptp\-format\fR
.IP
Equivalent to \fB\-\-tptp\-in\fR and \fB\-\-tptp\-out\fR.
.HP
\fB\-\-tptp2\-in\fR
.IP
Synonymous with \fB\-\-tptp\-in\fR.
.HP
\fB\-\-tptp2\-out\fR
.IP
Synonymous with \fB\-\-tptp\-out\fR.
.HP
\fB\-\-tptp2\-format\fR
.IP
Synonymous with \fB\-\-tptp\-format\fR.
.HP
\fB\-\-tstp\-in\fR
.IP
Parse TPTP\-3 format instead of E\-LOP (Note that TPTP\-3 syntax is still
under development, and the version in E may not be fully conforming at
all times. E works on all TPTP 3.0.1 input files (including includes).
.HP
\fB\-\-tstp\-out\fR
.IP
Print output clauses in TPTP\-3 syntax. In particular, for output levels
>=2, write derivations as TPTP\-3 derivations (default is PCL).
.HP
\fB\-\-tstp\-format\fR
.IP
Equivalent to \fB\-\-tstp\-in\fR and \fB\-\-tstp\-out\fR.
.HP
\fB\-\-tptp3\-in\fR
.IP
Synonymous with \fB\-\-tstp\-in\fR.
.HP
\fB\-\-tptp3\-out\fR
.IP
Synonymous with \fB\-\-tstp\-out\fR.
.HP
\fB\-\-tptp3\-format\fR
.IP
Synonymous with \fB\-\-tstp\-format\fR.
.HP
\fB\-\-auto\fR
.IP
Automatically determine settings for proof search. This is equivalent to
\fB\-xAuto\fR \fB\-tAuto\fR \fB\-\-sine\fR=\fIAuto\fR.
.HP
\fB\-\-satauto\fR
.IP
Automatically determine settings for proof/saturation search. This is
equivalent to \fB\-xAuto\fR \fB\-tAuto\fR.
.HP
\fB\-\-autodev\fR
.IP
Automatically determine settings for proof search (development version).
This is equivalent to \fB\-xAutoDev\fR \fB\-tAutoDev\fR \fB\-\-sine\fR=\fIAuto\fR.
.HP
\fB\-\-satautodev\fR
.IP
Automatically determine settings for proof/saturation search (development
version). This is equivalent to \fB\-xAutoDev\fR \fB\-tAutoDev\fR.
.HP
\fB\-\-auto\-schedule\fR
.IP
Use the (experimental) strategy scheduling. This will try several
different fully specified search strategies (aka "Auto\-Modes"), one after
the other, until a proof or saturation is found, or the time limit is
exceeded.
.HP
\fB\-\-no\-preprocessing\fR
.IP
Do not perform preprocessing on the initial clause set. Preprocessing
currently removes tautologies and orders terms, literals and clauses in a
certain ("canonical") way before anything else happens. Unless limited by
one of the following options, it will also unfold equational definitions.
.HP
\fB\-\-eq\-unfold\-limit=\fR<arg>
.IP
During preprocessing, limit unfolding (and removing) of equational
definitions to those where the expanded definition is at most the given
limit bigger (in terms of standard weight) than the defined term.
.HP
\fB\-\-eq\-unfold\-maxclauses=\fR<arg>
.IP
During preprocessing, don't try unfolding of equational definitions if
the problem has more than this limit of clauses.
.HP
\fB\-\-no\-eq\-unfolding\fR
.IP
During preprocessing, abstain from unfolding (and removing) equational
definitions.
.HP
\fB\-\-sine[=\fR<arg>]
.IP
Apply SInE to prune the unprocessed axioms with the specified filter.
\&'Auto' will automatically pick a filter. The option without the optional
argument is equivalent to \fB\-\-sine\fR=\fIAuto\fR.
.HP
\fB\-\-rel\-pruning\-level[=\fR<arg>]
.IP
Perform relevancy pruning up to the given level on the unprocessed
axioms. The option without the optional argument is equivalent to
\fB\-\-rel\-pruning\-level\fR=\fI3\fR.
.HP
\fB\-\-presat\-simplify\fR
.IP
Before proper saturation do a complete interreduction of the proof state.
.HP
\fB\-\-ac\-handling[=\fR<arg>]
.IP
Select AC handling mode. Preselected is 'DiscardAll', other options are
\&'None' to disable AC handling, 'KeepUnits', and 'KeepOrientable'. The
option without the optional argument is equivalent to
\fB\-\-ac\-handling\fR=\fIKeepUnits\fR.
.HP
\fB\-\-ac\-non\-aggressive\fR
.IP
Do AC resolution on negative literals only on processing (by default, AC
resolution is done after clause creation). Only effective if AC handling
is not disabled.
.HP
\fB\-W\fR <arg>
.HP
\fB\-\-literal\-selection\-strategy=\fR<arg>
.IP
Choose a strategy for selection of negative literals. There are two
special values for this option: NoSelection will select no literal (i.e.
perform normal superposition) and NoGeneration will inhibit all
generating inferences. For a list of the other (hopefully
self\-documenting) values run 'eprover \fB\-W\fR none'. There are two variants of
each strategy. The one prefixed with 'P' will allow paramodulation into
maximal positive literals in addition to paramodulation into maximal
selected negative literals.
.HP
\fB\-\-no\-generation\fR
.IP
Don't perform any generating inferences (equivalent to
\fB\-\-literal\-selection\-strategy\fR=\fINoGeneration\fR).
.HP
\fB\-\-select\-on\-processing\-only\fR
.IP
Perform literal selection at processing time only (i.e. select only in
the _given clause_), not before clause evaluation. This is relevant
because many clause selection heuristics give special consideration to
maximal or selected literals.
.HP
\fB\-i\fR
.HP
\fB\-\-inherit\-paramod\-literals\fR
.IP
Always select the negative literals a previous inference paramodulated
into (if possible). If no such literal exists, select as dictated by the
selection strategy.
.HP
\fB\-j\fR
.HP
\fB\-\-inherit\-goal\-pm\-literals\fR
.IP
In a goal (all negative clause), always select the negative literals a
previous inference paramodulated into (if possible). If no such literal
exists, select as dictated by the selection strategy.
.HP
\fB\-j\fR
.HP
\fB\-\-inherit\-conjecture\-pm\-literals\fR
.IP
In a conjecture\-derived clause), always select the negative literals a
previous inference paramodulated into (if possible). If no such literal
exists, select as dictated by the selection strategy.
.HP
\fB\-\-selection\-pos\-min=\fR<arg>
.IP
Set a lower limit for the number of positive literals a clause must have
to be eligible for literal selection.
.HP
\fB\-\-selection\-pos\-max=\fR<arg>
.IP
Set a upper limit for the number of positive literals a clause can have
to be eligible for literal selection.
.HP
\fB\-\-selection\-neg\-min=\fR<arg>
.IP
Set a lower limit for the number of negative literals a clause must have
to be eligible for literal selection.
.HP
\fB\-\-selection\-neg\-max=\fR<arg>
.IP
Set a upper limit for the number of negative literals a clause can have
to be eligible for literal selection.
.HP
\fB\-\-selection\-all\-min=\fR<arg>
.IP
Set a lower limit for the number of literals a clause must have to be
eligible for literal selection.
.HP
\fB\-\-selection\-all\-max=\fR<arg>
.IP
Set an upper limit for the number of literals a clause must have to be
eligible for literal selection.
.HP
\fB\-\-selection\-weight\-min=\fR<arg>
.IP
Set the minimum weight a clause must have to be eligible for literal
selection.
.HP
\fB\-\-prefer\-initial\-clauses\fR
.IP
Always process all initial clauses first.
.HP
\fB\-x\fR <arg>
.HP
\fB\-\-expert\-heuristic=\fR<arg>
.IP
Select one of the clause selection heuristics. Currently at least
available: Auto, Weight, StandardWeight, RWeight, FIFO, LIFO, Uniq,
UseWatchlist. For a full list check HEURISTICS/che_proofcontrol.c. Auto
is recommended if you only want to find a proof. It is special in that it
will also set some additional options. To have optimal performance, you
also should specify \fB\-tAuto\fR to select a good term ordering. LIFO is unfair
and will make the prover incomplete. Uniq is used internally and is not
very useful in most cases. You can define more heuristics using the
option \fB\-H\fR (see below).
.HP
\fB\-\-filter\-limit[=\fR<arg>]
.IP
Set the limit on the number of 'storage units' in the proof state, after
which the set of unprocessed clauses will be filtered against the
processed clauses to eliminate redundant clauses. As of E 0.7, a 'storage
unit' is approximately one byte, however, storage is estimated in an
abstract way, independent of hardware or memory allocation library, and
the storage estimate is only an approximation. The option without the
optional argument is equivalent to \fB\-\-filter\-limit\fR=\fI1000000\fR.
.HP
\fB\-\-filter\-copies\-limit[=\fR<arg>]
.IP
Set the number of storage units in new unprocessed clauses after which
the set of unprocessed clauses will be filtered for equivalent copies of
clauses (see above). As this operation is cheaper, you may want to set
this limit lower than \fB\-\-filter\-limit\fR. The option without the optional
argument is equivalent to \fB\-\-filter\-copies\-limit\fR=\fI800000\fR.
.HP
\fB\-\-delete\-bad\-limit[=\fR<arg>]
.IP
Set the number of storage units after which bad clauses are deleted
without further consideration. This causes the prover to be potentially
incomplete, but will allow you to limit the maximum amount of memory used
fairly well. The prover will tell you if a proof attempt failed due to
the incompleteness introduced by this option. It is recommended to set
this limit significantly higher than \fB\-\-filter\-limit\fR or
\fB\-\-filter\-copies\-limit\fR. If you select \fB\-xAuto\fR and set a memory limit, the
prover will determine a good value automatically. The option without the
optional argument is equivalent to \fB\-\-delete\-bad\-limit\fR=\fI1500000\fR.
.HP
\fB\-\-assume\-completeness\fR
.IP
There are various way (e.g. the next few options) to configure the prover
to be strongly incomplete in the general case. E will detect when such an
option is selected and return corresponding exit states (i.e. it will not
claim satisfiability just because it ran out of unprocessed clauses). If
you _know_ that for your class of problems the selected strategy is still
complete, use this option to tell the system that this is the case.
.HP
\fB\-\-assume\-incompleteness\fR
.IP
This option instructs the prover to assume incompleteness (typically
because the axiomatization already is incomplete because axioms have been
filtered before they are handed to the system.
.HP
\fB\-\-disable\-eq\-factoring\fR
.IP
Disable equality factoring. This makes the prover incomplete for general
non\-Horn problems, but helps for some specialized classes. It is not
necessary to disable equality factoring for Horn problems, as Horn
clauses are not factored anyways.
.HP
\fB\-\-disable\-paramod\-into\-neg\-units\fR
.IP
Disable paramodulation into negative unit clause. This makes the prover
incomplete in the general case, but helps for some specialized classes.
.HP
\fB\-\-condense\fR
.IP
Enable condensing for the given clause. Condensing replaces a clause by a
more general factor (if such a factor exists).
.HP
\fB\-\-condense\-aggressive\fR
.IP
Enable condensing for the given and newly generated clauses.
.HP
\fB\-\-disable\-given\-clause\-fw\-contraction\fR
.IP
Disable simplification and subsumption of the newly selected given clause
(clauses are still simplified when they are generated). In general, this
breaks some basic assumptions of the DISCOUNT loop proof search
procedure. However, there are some problem classes in which  this
simplifications empirically never occurs. In such cases, we can save
significant overhead. The option _should_ work in all cases, but is not
expected to improve things in most cases.
.HP
\fB\-\-simul\-paramod\fR
.IP
Use simultaneous paramodulation to implement superposition. Default is to
use plain paramodulation. This is an experimental feature.
.HP
\fB\-\-oriented\-simul\-paramod\fR
.IP
Use simultaneous paramodulation for oriented from\-literals. This is an
experimental feature.
.HP
\fB\-\-split\-clauses[=\fR<arg>]
.IP
Determine which clauses should be subject to splitting. The argument is
the binary 'OR' of values for the desired classes:
.TP
1:
Horn clauses
.TP
2:
Non\-Horn clauses
.TP
4:
Negative clauses
.TP
8:
Positive clauses
.TP
16:
Clauses with both positive and negative literals
.IP
Each set bit adds that class to the set of clauses which will be split.
The option without the optional argument is equivalent to
\fB\-\-split\-clauses\fR=\fI7\fR.
.HP
\fB\-\-split\-method=\fR<arg>
.IP
Determine how to treat ground literals in splitting. The argument is
either '0' to denote no splitting of ground literals (they are all
assigned to the first split clause produced), '1' to denote that all
ground literals should form a single new clause, or '2', in which case
ground literals are treated as usual and are all split off into
individual clauses.
.HP
\fB\-\-split\-aggressive\fR
.IP
Apply splitting to new clauses (after simplification) and before
evaluation. By default, splitting (if activated) is only performed on
selected clauses.
.HP
\fB\-\-split\-reuse\-defs\fR
.IP
If possible, reuse previous definitions for splitting.
.HP
\fB\-\-reweight\-limit[=\fR<arg>]
.IP
Set the number of new unprocessed clauses after which the set of
unprocessed clauses will be reevaluated. The option without the optional
argument is equivalent to \fB\-\-reweight\-limit\fR=\fI30000\fR.
.HP
\fB\-t\fR <arg>
.HP
\fB\-\-term\-ordering=\fR<arg>
.IP
Select an ordering type (currently Auto, LPO, LPO4, KBO or KBO6). \fB\-tAuto\fR
is suggested, in particular with \fB\-xAuto\fR. KBO and KBO1 are different
implementations of the same ordering, KBO is usually faster and has had
more testing. Similarly, LPO4 is an new, equivalent but superior
implementation of LPO.
.HP
\fB\-w\fR <arg>
.HP
\fB\-\-order\-weight\-generation=\fR<arg>
.IP
Select a method for the generation of weights for use with the term
ordering. Run 'eprover \fB\-w\fR none' for a list of options.
.HP
\fB\-\-order\-weights=\fR<arg>
.IP
Describe a (partial) assignments of weights to function symbols for term
orderings (in particular, KBO). You can specify a list of weights of the
form 'f1:w1,f2:w2, ...'. Since a total weight assignment is needed, E
will _first_ apply any weight generation scheme specified (or the default
one), and then modify the weights as specified. Note that E performs only
very basic sanity checks, so you probably can specify weights that break
KBO constraints.
.HP
\fB\-G\fR <arg>
.HP
\fB\-\-order\-precedence\-generation=\fR<arg>
.IP
Select a method for the generation of a precedence for use with the term

src/java/com/articulate/sigma/EProver.java
package com.articulate.sigma;
/** This code is copyright Articulate Software (c) 2014.  
This software is released under the GNU Public License <http://www.gnu.org/copyleft/gpl.html>.
Users of this code also consent, by use of this code, to credit Articulate Software
and Teknowledge in any writings, briefings, publications, presentations, or 
other representations of any software which incorporates, builds on, or uses this 
code.  Please cite the following article in any publication with references:

Pease, A., (2003). The Sigma Ontology Development Environment, 
in Working Notes of the IJCAI-2003 Workshop on Ontology and Distributed Systems,
August 9, Acapulco, Mexico.  See also sigmakee.sourceforge.net
*/

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
public class EProver {

    private Process _eprover;
    private BufferedReader _reader; 
    private BufferedWriter _writer; 
    private BufferedReader _error;
    private static String __dummyKBdir = KBmanager.getMgr().getPref("kbDir");

    /** *************************************************************
     *  */
    public static void writeBatchConfig(String inputFilename) {

        try {
            System.out.println("INFO in EProver.writeBatchFile(): writing EBatchConfig.txt with KB file " + inputFilename);
            //File initFile = new File(KBmanager.getMgr().getPref("kbDir"), "EBatchConfig.txt");
            File initFile = new File(__dummyKBdir, "EBatchConfig.txt");
            PrintWriter pw = new PrintWriter(initFile);
    
            pw.println("% SZS start BatchConfiguration");
            pw.println("division.category LTB.SMO");
            pw.println("output.required Assurance");
            pw.println("output.desired Proof Answer");
            pw.println("limit.time.problem.wc 60");
            pw.println("% SZS end BatchConfiguration");
            pw.println("% SZS start BatchIncludes");
            pw.println("include('" + inputFilename + "').");
            pw.println("% SZS end BatchIncludes");
            pw.println("% SZS start BatchProblems");
            pw.println("% SZS end BatchProblems");
            pw.close();
        }
        catch (Exception e1) {
            e1.printStackTrace();
            System.out.println("Error in EProver.writeBatchFile()");
            System.out.println(e1.getMessage());
        }
    }
    
    /** *************************************************************
     */
    private EProver () {
    }

    /** *************************************************************
     * Creates a running instance of EProver.  To obtain a new
     * instance of EProver, use the static factory method
     * EProver.getNewInstance().
     *
     * @param executable A File object denoting the platform-specific
     * EProver executable.
     *
     * @param kbFile A File object denoting the initial knowledge base
     * to be loaded by the EProver executable.
     *
     * @throws IOException should not normally be thrown unless either
     *         EProver executable or database file name are incorrect
     *         
     * e_ltb_runner --interactive LTBSampleInput-AP.txt
     */
    public EProver (String executable, String kbFile) throws IOException {

        writeBatchConfig(kbFile);
        System.out.println("INFO in EProver(): executable: " + executable);
        System.out.println("INFO in EProver(): kbFile: " + kbFile);
        // String execString = executable + " --interactive " + 
                // KBmanager.getMgr().getPref("kbDir") + File.separator + "EBatchConfig.txt";
        // Qingqing: modify executable string using e_ltb_runner
        String execString = executable + " --interactive "
                + __dummyKBdir + File.separator + "EBatchConfig.txt "
                + executable.substring(0, executable.lastIndexOf("/")) + File.separator + "eprover "
                + executable.substring(0, executable.lastIndexOf("/")) + File.separator + "epclextract";
        System.out.println("INFO in EProver(): executing: " + execString);
        _eprover = Runtime.getRuntime().exec(execString);
        _reader = new BufferedReader(new InputStreamReader(_eprover.getInputStream()));
        _error = new BufferedReader(new InputStreamReader(_eprover.getErrorStream()));
        System.out.println("INFO in EProver(): initializing process");

        // Qingqing: another way to write while-loop to avoid NullPointerException
        String line = _reader.readLine();
        while (line != null) {
            if (line.indexOf("Error:") != -1)
                throw new IOException(line);
            if (line.indexOf("# Enter job") != -1)
                break;
            line = _reader.readLine();
        }
        _writer = new BufferedWriter(new OutputStreamWriter(_eprover.getOutputStream()));
    }
    
    /** *************************************************************
     * Add an assertion.
     *
     * @param formula asserted formula in the KIF syntax
     * @return answer to the assertion (in the XML syntax)
     * @throws IOException should not normally be thrown
     */
    public String assertFormula(String formula) {
        //public String assertFormula(String formula) throws IOException {

        String result = "";
        try {
            // String assertion = SUMOformulaToTPTPformula.tptpParseSUOKIFString(formula,false);
            String assertion = "";
            _writer.write(assertion);
            _writer.flush();
            String line;
            do {
                line = _reader.readLine();  
                if (line.indexOf("Error:") != -1) 
                    throw new IOException(line);                
                System.out.println("INFO EProver(): Response: " + line);
                result += line + "\n";
                if (line.indexOf("# Processing finished") != -1) 
                    break;                
            } while (line != null);
        }
        catch (Exception ex) {
            System.out.println("Error in EProver.assertFormula(" + formula + ")");
            System.out.println(ex.getMessage());
            ex.printStackTrace();
        }
        return result;
    }

    /** *************************************************************
     * Terminate this instance of EProver. 
     * <font color='red'><b>Warning:</b></font>After calling this function
     * no further assertions or queries can be done.
     *
     * @throws IOException should not normally be thrown
     */
    public void terminate() throws IOException {

        System.out.println();
        System.out.println("TERMINATING " + this);
        try {
            _writer.write("quit.\n");
            _writer.write("go.\n");
            _writer.close();
            _reader.close();
            System.out.println("DESTROYING the Process " + _eprover);
            System.out.println();
            _eprover.destroy();
        }
        catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    /** *************************************************************
     * Submit a query.
     *
     * @param formula query in the KIF syntax
     * @param kb current knowledge base
     * @return answer to the query 
     * @throws IOException should not normally be thrown
     */
    public String submitQuery(String formula, KB kb) {
        //public String submitQuery (String formula, int timeLimit, int bindingsLimit) throws IOException {
                
        String result = "";
        System.out.println("INFO in EProver.submitQuery() formula: " + formula);
        //Formula f = new Formula();
        //f.read(formula);
        //SUMOformulaToTPTPformula sfttptp = new SUMOformulaToTPTPformula();

        try {
            //ArrayList<String> al = sfttptp.tptpParse(f, true, kb);
            String query = SUMOformulaToTPTPformula.tptpParseSUOKIFString(formula,true);
            System.out.println("INFO in EProver.submitQuery() TPTP formula: " + query);
            //System.out.println("INFO in EProver.submitQuery() TPTP formula: " + al.get(0));
            //String conjecture = "fof(conj1,conjecture, " + al.get(0) + ").";
            String conjecture = "fof(conj1,conjecture, " + query + ").";
            //String tptpAssert = "( ( ? [V__X] : s__subclass(V__X,s__Object) ) )";
            //System.out.println("INFO in EProver.submitQuery() TPTP formula: " + tptpAssert);
            //String conjecture = "fof(conj1,conjecture, " + tptpAssert + ").";
            System.out.println("INFO in EProver.submitQuery() conjecture: " + conjecture);
            _writer.write(conjecture + "\n");
            _writer.write("go.\n");
            _writer.flush();

            // Qingqing: another way to write while-loop to avoid NullPointerException
            //System.out.println("INFO in EProver.submitQuery() writing executing results.");
            String line = _reader.readLine();
            boolean inProof = false;
            while (line != null) {
                if (line.indexOf("# Enter job") != -1)
                    break;
                if (line.indexOf("# SZS status") != -1)
                    inProof = true;
                if (inProof)
                    result += line + "\n";

                line = _reader.readLine();
            }
        }
        catch (Exception ex) {
            System.out.println("Error in EProver.submitQuery(): " + ex.getMessage());
            ex.printStackTrace();
        }
        return result;
    }

    /** *************************************************************
     * A simple test. Works as follows: 
     * <ol>
     *   <li>start E;</li>
     *   <li>make an assertion;</li>
     *   <li>submit a query;</li>
     *   <li>terminate E</li>
     *</ol>
     */
    public static void main(String[] args) throws Exception {

        /*
        String initialDatabase = "SUMO-v.kif";
        EProver eprover = EProver.getNewInstance(initialDatabase);
        eprover.setCommandLineOptions("--cpu-limit=600 --soft-cpu-limit=500 -xAuto -tAuto -l 4 --tptp3-in");
        KBmanager.getMgr().setPref("eprover","/home/apease/Programs/E/Prover/eprover");
        System.out.print(eprover.submitQuery("(holds instance ?X Relation)",5,2));
*/
        try {
            System.out.println("INFO in EProver.main()");
            //KBmanager.getMgr().initializeOnce();
            //KB kb = KBmanager.getMgr().getKB("SUMO");
            KB kb = null;
            System.out.println("------------- INFO in EProver.main() completed initialization--------");
            EProver eprover = new EProver("/home/apease/Programs/E/PROVER/e_ltb_runner",
                    "/home/apease/Sigma/KBs/SUMO.tptp");

            System.out.println(eprover.submitQuery("(subclass ?X Object)",kb));
            eprover.terminate();
        } 
        catch (Exception e) {
            System.out.println(e.getMessage());
        }

        // System.out.print(eprover.assertFormula("(human Socrates)"));
        // System.out.print(eprover.assertFormula("(holds instance Adam Human)"));
        // System.out.print(eprover.submitQuery("(human ?X)", 1, 2));
        // System.out.print(eprover.submitQuery("(holds instance ?X Human)", 5, 2));
        
    }
    
}

